### ISSUE TITLE
Lua tile-material module fails on most tiles

### ISSUE THREAD

**Issue by lethosor** [DFHack/dfhack#1956]
Reported by teatime on [Discord](https://discord.com/channels/793331351645323264/807444467140788254/883793343887319090):

> There exists a `tile-material.lua`, and has apparently since 2015. Operation is claimed to just be `GetTileMat(x,y,z)` but this seems to always give me nil, even when there is a valid tile at that position (as tested wtih a more specific function from the same file).
I've narrowed the issue to `tile-material.lua:356`, but I come from a faraway land, the customs are strange to me, and i don't really communicate well in the local language; if someone would care to help, the relevant information seems to be:
> 
> - I have a mineral tile, it exists at a pos, where `local pos = {x  = ???, y = ???, z = ???}`
> - `df.tiletype.attrs[dfhack.maps.getTileType(pos)].material` is used to get a numerical representation of the tile type, in this case, a healthy 5
> - The 5 brings me joy, for it corresponds with `df.tiletype_material.MINERAL`, as expected.
> - There exists a table in `tile-material.lua`, which contains the following:
> ```lua
> BasicMats = {
>     --some stuff
>     [df.tiletype_material.MINERAL] = GetVeinMat,
> ```
> ..which references a function, function `GetVeinMat(x, y, z)`
> - This function works; see how it works:
> ```lua
> local mat = tilemat.GetVeinMat(pos)
> print("lookatme! "..tostring(mat)) -- prints: lookatme! <material 0:196 INORGANIC:COAL_BITUMINOUS>
> ```
> 
> So far so good - but alas, here things go wrong.
> - Once one enters the `GetTileMat(x,y,z)` function, which abstracts over all materials, one is propelled dowards, bringing along the BasicMats, until finally the following code is hit:
>     ```lua
>     local mat_getter = matspec[type_mat]
>     if mat_getter == nil then
>         return nil
>     end
>     return mat_getter(pos)
>     ```
> 
> `matspec` is `BasicMat`, the one that transforms the blessed 5 into `GetVeinMat`, but alas, it always falls through as `nil`.
> BasicMat breaks what appears to be the contract - `BasicMat[df.tiletype_material.MINERAL]` curses me with a `nil`, when it seems it should've given me a function.

@lethosor:

> I think that module is broken
> ```
> [lua]# m=require('tile-material')
> [lua]# ~m.BasicMats
> table: 0x7fffbc147500
> 6                           = function: 0x7fffbc147110
> ```
> because of lua scoping rules, the functions don't actually exist at the time `BasicMats` is initialized, so nearly every entry is nil

**Comment by lethosor** [comment]
Two possible solutions:

- Move the `BasicMats` table declaration (and others, most likely) to the end of the file so that they are initialized properly.
- Remove the module, and expose the API implicitly requested in #1031 to Lua.
  - This is unlikely to cause many problems because given how broken the module is, it is unlikely to be used. If it is used, we could hopefully add a stub for compatibility fairly easily.

**Comment by milochristiansen** [comment]
I'm not sure how this module is broken since I'm 90% sure I used it in at least some of my mods back in the day (which is why I made it in the first place), but looking at the code. Yup. It sure is broken all right. And in such a stupid way too.

I was notified about this bug because someone emailed me about the license on the code, so I'll just reiterate what I told them here:

Anyone modifying anything I ever added to DFHack has two options specifically blessed by me (although feel free to do anything else you want allowed by the license, I don't actually care much at all):

* Just add their name to the copyright notice in addition to mine.
* Change the license header to CC0 with the following note `Original code provided by Milo Christiansen in <year> under the <name of original license, probably MIT> license. Relicensed for ease of maintenance with his permission.` (obviously this note isn't required by CC0, but I would appreciate it).
